tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/AspNetCore/AspNetCoreBlockMiddlewareIntegration.cs

```diff
@@ -0,0 +1,109 @@
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Microsoft.AspNetCore.Hosting",
+        TypeName = "Microsoft.AspNetCore.Hosting.Builder.ApplicationBuilderFactory",
+        MethodName = "CreateBuilder",
+        ParameterTypeNames = new[] { "Microsoft.AspNetCore.Http.Features.IFeatureCollection" },
+        ReturnTypeName = "Microsoft.AspNetCore.Builder.IApplicationBuilder",
+        MinimumVersion = "2",
+        MaximumVersion = "6",
+        IntegrationName = nameof(IntegrationId.AspNetCore))]
+    public class AspNetCoreBlockMiddlewareIntegration
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(AspNetCoreBlockMiddlewareIntegration));
+
+        /// <summary>
+        /// test
+        /// </summary>
+        /// <param name="instance">instance</param>
+        /// <param name="returnValue">returnValue</param>
+        /// <param name="exception">exception</param>
+        /// <param name="state">state</param>
+        /// <typeparam name="TTarget">TTarget</typeparam>
+        /// <typeparam name="TReturn">TReturn</typeparam>
+        /// <returns>CallTargetReturn</returns>
+        public static CallTargetReturn<TReturn> OnMethodEnd<TTarget, TReturn>(TTarget instance, TReturn returnValue, Exception exception, in CallTargetState state)
+        {
+#if !NETFRAMEWORK
+            if (Security.Instance.Settings.Enabled)
+            {
+                var appb = (IApplicationBuilder)returnValue;
+                appb.MapWhen(context => context.Items["block"] is true, HandleBranch);
+            }
+#endif
+
+            return new CallTargetReturn<TReturn>(returnValue);
+        }
+
+#if !NETFRAMEWORK
+        internal static void HandleBranch(Microsoft.AspNetCore.Builder.IApplicationBuilder app)
+        {
+            app.Run(
+                context =>
+                {
+                    var sec = Security.Instance;
+                    var settings = sec.Settings;
+                    var httpResponse = context.Response;
+                    httpResponse.Clear();
+
+                    foreach (var cookie in context.Request.Cookies)
+                    {
+                        httpResponse.Cookies.Delete(cookie.Key);
+                    }
+
+                    // this should always be true for core, but it would seem foolish to ignore it, as potential source of future bugs
+                    if (sec.CanAccessHeaders())
+                    {
+                        httpResponse.Headers.Clear();
+                    }
+
+                    httpResponse.StatusCode = 403;
+                    var syncIOFeature = context.Features.Get<IHttpBodyControlFeature>();
+                    if (syncIOFeature != null)
+                    {
+                        // allow synchronous operations for net core >=3.1 otherwise invalidoperation exception
+                        syncIOFeature.AllowSynchronousIO = true;
+                    }
```

Comment:
I think you can remove all this now that you're using `httpResponse.WriteAsync()`? 🎉 
```suggestion
```